### PR TITLE
fix: report on correct location instead of entire quasi

### PR DESCRIPTION
**Pull request by thepassle**
before:
Screenshot 2020-10-16 at 19 54 30

after:
Screenshot 2020-10-16 at 19 48 10

### CODE REVIEW

**Comment by thepassle** [comment]
Thinking about this more, I dont think the logic to 'select' the quasi is even needed anymore?

**Comment by thepassle** [comment]
Yep, seems like it's not needed any more. Checked with a few more test cases to make sure everything still works: 
image

@43081j PTAL 🙂

**Comment by 43081j** [comment]
i think there might be an edge case here with the column.

can you try:

```ts
render() {
  return html`<div attr="<">
    Contents
  </div>`;
}
```

as i imagine using `endCol - 1` for every line _except the first_ will work. since the first line's column would actually be `(colOfFirstQuasi + endCol)`.

as in:

```ts
render() {
  return html`<div attr="<">
--------------^-- this is (14 + 0) i suppose
    Content
---^-- this is just 4, in both estree and parse5
  </div>`;
}
```

**Comment by thepassle** [comment]
Good spot, looks like that does report incorrectly:
image

I'll take another look today

**Comment by 43081j** [comment]
haha, i did just check the AST of both and yep, one is 0-based and the other is 1-based. consistent 😂 

i also had another thought just now which is a little unfortunate... what if we have something like this:

```ts
render() {
  return html`
    Foo ${someVariable} Bar
  `;
}
```

The `${someVariable}` will be a different length to the placeholder we internally add to the DOM.

So it may be the case that we do actually need to use the quasi's column rather than just the node's. not entirely sure off the top of my head how we'd solve that yet but you can recreate it probably by:

```ts
render() {
  return html`
    <div bar=${whatever} attr="<"></div>
  `;
}
```

because in reality the parse5 tree will be like:

```html
<div bar="{{SOME_PLACEHOLDER}}" attr="<"></div>
```

the col of `attr` would differ

**Comment by thepassle** [comment]
> So it may be the case that we do actually need to use the quasi's column rather than just the node's. not entirely sure off the top of my head how we'd solve that yet but you can recreate it probably by:

Could this be fixed if we make the placeholder the same length as the expression? e.g.:

```html
<div bar=${foobar}></div>
```

```
{{12345}}
${foobar}
```

**Comment by thepassle** [comment]
First edgecase is fixed:

Screenshot 2020-10-17 at 13 48 47

Looking into the second one now

**Comment by thepassle** [comment]
Hrm, I got _something_ working seemingly correctly for edgecase 2, but its not the prettiest code 🙈 Could you take a look @43081j ?

Here's a screenshot:

Screenshot 2020-10-17 at 14 37 02

**Comment by thepassle** [comment]
Changing the placeholder will break the `isExpressionPlaceholder` check, but I dont see that used anywhere? 🤔

**Comment by 43081j** [comment]
it would be nice to maintain a single placeholder. can we maybe use the quasi's line/col as offset? if we know that, we know the line and column the "string" begins at. we can locate the placeholder like in master and start from there maybe.

i can have a look tomorrow too if we're still stuck then

**Comment by thepassle** [comment]
Looks like more edgecases ☹️ Think I'll need some assistance after all 😓 

correct:
Screenshot 2020-10-17 at 23 31 00

but here it reports the error on the `}` after the `.map`
Screenshot 2020-10-17 at 23 30 48

**Comment by thepassle** [comment]
I've updated the PR

File: src/template-analyzer.ts [hunk 1]
```diff
@@ -198,19 +198,40 @@ export class TemplateAnalyzer {
...
+    let expression;
+    let height = 0;
 
     for (const quasi of this._node.quasi.quasis) {
       const placeholder = getExpressionPlaceholder(this._node, quasi);
       offset += quasi.value.raw.length + placeholder.length;
 
+      const i = this._node.quasi.quasis.indexOf(quasi);
+      if (i !== 0) {
+        expression = this._node.quasi.expressions[i - 1];
```
Comment (43081j) [inline src/template-analyzer.ts:212]: i think this can be scoped now as you only use it here. as in:

```ts
const expression \ this._node.quasi.expressions[i - 1];
```

and get rid of the `let expression;` further up

File: src/util.ts [hunk 2]
```diff
@@ -107,6 +107,7 @@ export function getExpressionPlaceholder(
   return `{{__Q:${i}__}}`;
 }
 
+
```
Comment (43081j) [inline src/util.ts:110]: what happened here? 🤔

Comment (thepassle) [reply src/util.ts:110]: hum Im not sure o.O looks like I added a rogue line by accident

Comment (thepassle) [reply src/util.ts]: fixed :+1:

File: src/template-analyzer.ts (hunk 1 above)
Comment (thepassle) [reply src/template-analyzer.ts]: fixed 👍

**Comment by 43081j** [comment]
i suspect this doesn't build or tests fail but travis has decided to go walkies...

ill see whats up with it

File: src/template-analyzer.ts [hunk 3]
```diff
@@ -198,19 +198,39 @@ export class TemplateAnalyzer {
...
+    let height = 0;
 
     for (const quasi of this._node.quasi.quasis) {
       const placeholder = getExpressionPlaceholder(this._node, quasi);
       offset += quasi.value.raw.length + placeholder.length;
 
+      const i = this._node.quasi.quasis.indexOf(quasi);
+      if (i !== 0) {
+        const expression = this._node.quasi.expressions[i - 1];
+        height += expression.loc.end.line - expression.loc.start.line;
```
Comment (43081j) [inline src/template-analyzer.ts:212]: since `expression.loc` is nullable, we should return `null` i think and reintroduce that in the return type. like a short circuit

File: src/template-analyzer.ts [hunk 4]
```diff
@@ -198,19 +198,39 @@ export class TemplateAnalyzer {
    */
   public resolveLocation(
     loc: parse5.Location
-  ): ESTree.SourceLocation | null | undefined {
+  ): ESTree.SourceLocation {
     let offset = 0;
+    let height = 0;
 
     for (const quasi of this._node.quasi.quasis) {
```
Comment (43081j) [inline src/template-analyzer.ts:205]: above this, check if `this._node.loc` or `this._node.quasi.loc` is empty and, if it is, return `null`

**Comment by thepassle** [comment]
@43081j PTAL

**Comment by thepassle** [comment]
hrm looks like i broke a bunch of tests -.-

**Comment by thepassle** [comment]
okay, should be good to go now 👍

**Review by 43081j — APPROVED** [review]
